nemo/collections/asr/modules/flashlight_decoder.py

```diff
@@ -200,12 +200,10 @@ def __init__(
             )
         else:
             from flashlight.lib.text.decoder import LexiconFreeDecoder, LexiconFreeDecoderOptions
+            from flashlight.lib.text.dictionary import Dictionary
 
-            d = {
-                w: [[w]]
-                for w in self.tokenizer_wrapper.vocab + ([] if '<unk>' in self.tokenizer_wrapper.vocab else ['<unk>'])
-            }
-            self.word_dict = create_word_dict(d)
+            # Dictionary contains a mapping of our ASR model's output IDs to the tokens in the vocabulary
+            self.word_dict = Dictionary(self.tokenizer_wrapper.vocab)
```

Comment:
The dictionary has to contain the same number of IDs as the size of the ASR's output layer. I think we can't just add an `<unk>` token if it isn't already in the tokeniser's vocab.

Reply:
I would need a review for this one, but I'm heading to a conference. @nithinraok can you ping the author of the file for review

Reply:
@trias702

Reply:
Actually, if you remove the '<unk>' it causes a lot of problems with silences. I remember I added this hack because we had a lot of problems with lexicon-free decoding in that we were not generating spaces correctly, but adding this in fixed it. Although this is from 2 years ago, so my brain is a bit hazy with specifics, but I would strongly urge that you test out your new code without the '<unk>' in lexicon-free mode on English and Japanese to ensure you're not seeing any strange behaviour (it is also possible that whatever issue I was seeing 2 years ago when I added this hack has since been fixed in the C++ part of the flashlight code).

Reply:
@trias702 
Sorry for the late reply. I was at ICASSP and last week was Golden Week in Japan (default holiday season).

First things first, I haven't tested the implementation with a model/tokeniser that does not have an OOV token. All tokenisers that I'm using have an OOV token. I will have a look at the CTC probs for the OOV token, if it actually is output anywhere.

Regarding the Flashlight code, the corresponding decoding part is here
https://github.com/flashlight/text/blob/bbe9e3c201f5c9c3f3c0d553f0ea73af5e0a5209/flashlight/lib/text/decoder/LexiconFreeDecoder.cpp#L30-L125
The decoder iterates over the top `token_size` probabilities. If the AM doesn't have an OOV token, it should show up in the emission probabilities. Equally, if the LM doesn't have an OOV token, we should not be able to get a score for the OOV token, i.e., these lines should fail.
https://github.com/flashlight/text/blob/bbe9e3c201f5c9c3f3c0d553f0ea73af5e0a5209/flashlight/lib/text/decoder/LexiconFreeDecoder.cpp#L72-L73
because we get this error
https://github.com/flashlight/text/blob/bbe9e3c201f5c9c3f3c0d553f0ea73af5e0a5209/flashlight/lib/text/decoder/lm/KenLM.cpp#L66-L69

(I assume that an OOV token should be added after the tokeniser's vocab and have an index > `usrToLmIdxMap_.size()`)